### PR TITLE
Inital draft of hub documentation with examples

### PR DESCRIPTION
**Pull request by admd**

### CODE REVIEW

**Comment by admd** [comment]
@Loquacity  I can't add one as a reviewer. Please have a look.
@moio @chiaradiamarcelo

**Comment by Loquacity** [comment]
Hi @admd! Are you adding a new guide to our suite?

**Comment by admd** [comment]
Hi @Loquacity , I am sorry, should have added some description. Yes, we need to add a new section in the our documentation for the Hub. That guide then will consist of multiple subsections. For now, we are adding documentation for XML-RPC API section.

File: modules/hub/pages/hub-intro.adoc [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+
+Hub is an umbrella term. It is collection of number of components and improvement which could help
+users with large setups to easily manage their systems. A general problem statement is given below which was the main motivation behind this effort.
+
+
+===  Problem statement
+
+
+====  Scalability
+Some of our users need to manage large amounts of clients (tens of thousands). Currently, this is impossible to achieve using a single Server and there is no forecast of being able to accomplish that in the foreseeable future. We want to define a new SUSE Manager-based solution to provide a reasonable subset of SUSE Manager functionality at that scale.
```
Comment (chiaradiamarcelo) [inline modules/hub/pages/hub-intro.adoc]: Shouldn't we use "Uyuni" instead of SUSE Manager?

File: modules/hub/pages/hub-intro.adoc [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+
+====  Store independence
+An additional specific problem arises in retail environments, where the need is to manage a very large number of clients in a large number of physical locations (stores). Stores should be able to use a reasonable subset of SUSE Manager functionality locally even in temporary or prolonged absence of a connection to the Internet or a company intranet, in particular to a central data center.
```
Comment (chiaradiamarcelo) [inline modules/hub/pages/hub-intro.adoc]: Same as above

File: modules/hub/pages/relay-mode.adoc [hunk 3]
```diff
@@ -0,0 +1,54 @@
+[[manual-mode-example]]
+= Manual mode
```
Comment (chiaradiamarcelo) [inline modules/hub/pages/relay-mode.adoc]: Relay mode

File: modules/hub/pages/hub-api-usage.adoc [hunk 4]
```diff
@@ -0,0 +1,7 @@
+[[hub-api-namespaces]]
+= Hub XMRPC API Usage
+
+hub-xmlrpc-api is a daemon, currently to be launched from the command line.
```
Comment (chiaradiamarcelo) [inline modules/hub/pages/hub-api-usage.adoc]: I think if we are implementing a systemd solution we should change this

File: modules/hub/pages/hub-api-installation.adoc [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+
+Configuration of `hub-xmlrpc-api` is specified in a JSON file like the following:
+[source,json]
+----
+{
+   "type": "json",
+    "hub": {
+       "manager_api_url": "http://localhost/rpc/api"
+   },
+    "connect_timeout": 10,
```
Comment (chiaradiamarcelo) [inline modules/hub/pages/hub-api-installation.adoc]: I know it's quite intuitive, but shouldn't we add a small explanation of the timeouts anyway?

File: modules/hub/pages/manual-mode.adoc [hunk 6]
```diff
@@ -0,0 +1,56 @@
+[[manual-mode-example]]
+= Manual mode
+
+
+In this example, manual mode is used, where user needs to provide credentials for each {productname} server user wants to connect to.
+
+Here in this example, following is happening
+
+[square]
+* User provide credentials for the hub to the login method, a `sessionkey` for the hub is returned
```
Comment (chiaradiamarcelo) [inline modules/hub/pages/manual-mode.adoc]: provides

File: modules/hub/pages/manual-mode.adoc [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+* Using the sessionkey from the previous step, `serverIds` for the servers attached to the hub are requested
+* User provide the credentials for each {productname} server, she needs data from and calls the hub method `attachToServers`. Authentication part is done here.
```
Comment (chiaradiamarcelo) [inline modules/hub/pages/manual-mode.adoc]: provides

File: modules/hub/pages/manual-mode.adoc [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+= Manual mode
+
+
+In this example, manual mode is used, where user needs to provide credentials for each {productname} server user wants to connect to.
+
+Here in this example, following is happening
+
+[square]
+* User provide credentials for the hub to the login method, a `sessionkey` for the hub is returned
+* Using the sessionkey from the previous step, `serverIds` for the servers attached to the hub are requested
```
Comment (chiaradiamarcelo) [inline modules/hub/pages/manual-mode.adoc]: I would add "by calling 'listServerIds' or somehting like that

File: modules/hub/pages/manual-mode.adoc [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+* User provide the credentials for each {productname} server, she needs data from and calls the hub method `attachToServers`. Authentication part is done here.
+* `multicast` call is being performed on the specific servers where `serverId` is used to identify the specific sever. Behind the scenes, `system.list_system method` of each {productname} server's xmlrpc api method will be called to get the list of the clients registed with each {productname} server
```
Comment (chiaradiamarcelo) [inline modules/hub/pages/manual-mode.adoc]: I think the method is called 'system.listSystem'

Comment (chiaradiamarcelo) [inline modules/hub/pages/manual-mode.adoc]: a `multicast` call is performed on the specified servers.

Comment (chiaradiamarcelo) [inline modules/hub/pages/manual-mode.adoc]: some times 'xmlrpc api' is lower case and some others not. I think we should use the same way everywhere

Comment (chiaradiamarcelo) [inline modules/hub/pages/manual-mode.adoc]: typo in registered.

File: modules/hub/pages/manual-mode.adoc [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+* Hub will take care of collecting these results and return in form of a `map`. This map has two entries, entry with `Successfull`  key, has list of responses where call was successful while with `Failed` key  has the failed ones.
```
Comment (chiaradiamarcelo) [inline modules/hub/pages/manual-mode.adoc]: I would reword it to use 'aggregating' instead of 'collecting'.

The Hub will aggregate the results and return a response in form of a `map`, which has two entries:
`Successful`: contains a list of responses for those clients where the call was successful.
`Failed`: contains a list of responses  for those clients where the call failed.

File: modules/hub/pages/manual-mode.adoc [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+
+[NOTE]
+====
+If user wish to get information from only one {productname} server, then Hub also provided a convenient `unicast` namespace. In this case, result would not be returned as map rather exactly as the result are expectd as if user directly call that {productname} server's api.
```
Comment (chiaradiamarcelo) [inline modules/hub/pages/manual-mode.adoc]: If the user wish to get information from only one {productname} server, the Hub API also provides a convenient `unicast` namespace. In this case, the response would not be a `map`, instead the response format would be exactly the same as if the user would directly call that {productname} server's API.

File: modules/hub/pages/relay-mode.adoc [hunk 12]
```diff
@@ -0,0 +1,54 @@
...
+
+
+In this example, relay mode is used, this is a convenient method where it is assumed that each server has an admin user with the same credentials as one of Hub's user but user still needs to call `attachToServers` method.
```
Comment (chiaradiamarcelo) [inline modules/hub/pages/relay-mode.adoc]: I would reword it as something like:

The `Relay mode` is a convenient mode where the same credentials used to login into the Hub API will be used to login into the server's APIs as well the user wants to work with. It is assumed that the credentials are valid in each server and correspond to an admin user. After logging in into the Hub API the user still needs to call `attachToServers` method.

File: modules/hub/pages/relay-mode.adoc [hunk 13]
```diff
@@ -0,0 +1,54 @@
...
+Here in this example, following is happening
+
+[square]
+* User provide credentials for the hub to the login method, a `sessionkey` for the hub is returned
```
Comment (chiaradiamarcelo) [inline modules/hub/pages/relay-mode.adoc]: typo: provides

File: modules/hub/pages/relay-mode.adoc [hunk 14]
```diff
@@ -0,0 +1,54 @@
...
+* Using the sessionkey from the previous step, `serverIds` for the servers attached to the hub are requested
+* User calls the hub method `attachToServers`. Hub will use the same credentials used in step-1 to authenticate the specified servers.
```
Comment (chiaradiamarcelo) [inline modules/hub/pages/relay-mode.adoc]: 'step-1' -> in the first step

File: modules/hub/pages/relay-mode.adoc [hunk 15]
```diff
@@ -0,0 +1,54 @@
...
+* `multicast` call is being performed on the specific servers where `serverId` is used to identify the specific sever. Behind the scenes, `system.list_system method` of each {productname} server's xmlrpc api method will be called to get the list of the clients registed with each {productname} server
+* Hub will take care of collecting these results and return in form of a `map`. This map has two entries, entry with `Successfull`  key, has list of responses where call was successful while with `Failed` key  has the failed ones.
+
+[NOTE]
+====
+If user wish to get information from only one {productname} server, then Hub also provided a convenient `unicast` namespace. In this case, result would not be returned as map rather exactly as the result are expectd as if user directly call that {productname} server's api.
```
Comment (chiaradiamarcelo) [inline modules/hub/pages/relay-mode.adoc]: same as above https://github.com/uyuni-project/uyuni-docs/pull/166/files#r397190306

File: modules/hub/pages/relay-mode.adoc [hunk 16]
```diff
@@ -0,0 +1,54 @@
...
+
+
+In this example, relay mode is used, this is a convenient method where it is assumed that each server has an admin user with the same credentials as one of Hub's user but user still needs to call `attachToServers` method.
+Here in this example, following is happening
+
+[square]
+* User provide credentials for the hub to the login method, a `sessionkey` for the hub is returned
+* Using the sessionkey from the previous step, `serverIds` for the servers attached to the hub are requested
+* User calls the hub method `attachToServers`. Hub will use the same credentials used in step-1 to authenticate the specified servers.
+* `multicast` call is being performed on the specific servers where `serverId` is used to identify the specific sever. Behind the scenes, `system.list_system method` of each {productname} server's xmlrpc api method will be called to get the list of the clients registed with each {productname} server
```
Comment (chiaradiamarcelo) [inline modules/hub/pages/relay-mode.adoc]: same comments as in manual mode

File: modules/hub/pages/manual-mode.adoc [hunk 17]
```diff
@@ -0,0 +1,56 @@
+[[manual-mode-example]]
+= Manual mode
+
+
+In this example, manual mode is used, where user needs to provide credentials for each {productname} server user wants to connect to.
```
Comment (chiaradiamarcelo) [inline modules/hub/pages/manual-mode.adoc]: I would reword it as something like:

In the `Manual mode`  the user needs to provide credentials for each {productname} server the user wants to connect to.

File: modules/hub/pages/autoconnect-mode.adoc [hunk 18]
```diff
@@ -0,0 +1,51 @@
+[[manual-mode-example]]
+= Manual mode
+
+
+In this example, autoconnect mode is used, this is a convenient method where it is assumed that each server has an admin user with the same credentials as one of Hub's user.
```
Comment (chiaradiamarcelo) [inline modules/hub/pages/autoconnect-mode.adoc]: I would reword it as something like:

The `autoconnect` mode is a convenient method where the same credentials used to login into the Hub API are used to login into the APIs of the servers the user wants to work with. It is assumed the credentials are valid in each server and correspond to an admin user.

File: modules/hub/pages/autoconnect-mode.adoc [hunk 19]
```diff
@@ -0,0 +1,51 @@
...
+Difference from relay mode is that there is no need to call `attachToServers` method which also means that there is no need to make a call to get the `serverIds`. This process will happen automatically
+when user will login into hub using `loginWithAutoconnectMode` method. Selected servers will be the servers where Hub user will have access though.
+
+Here in this example, following is happening
+
+[square]
+* User provide credentials for the hub to the `loginWithAutoconnectMode` method, a `sessionkey` for the hub is returned. 
+* `multicast` call is being perfomed on the specific servers where `serverId` is used to indentify the specific sever. Behind the scenes, `system.list_system method` of each {productname} server's xmlrpc api method will be called to get the list of the clients registed with each {productname} server
```
Comment (chiaradiamarcelo) [inline modules/hub/pages/autoconnect-mode.adoc]: same comments as in manual mode example

File: modules/hub/pages/autoconnect-mode.adoc [hunk 20]
```diff
@@ -0,0 +1,51 @@
...
+* Hub will take care of collecting these results and return in form of a `map`. This map has two entries, entry with `Successfull`  key, has list of responses where call was successful while with `Failed` key  has the failed ones.
+
+[NOTE]
+====
+If user wish to get information from only one {productname} server, then Hub also provided a convenient `unicast` namespace. In this case, result would not be returned as map rather exactly as the result are expectd as if user directly call that {productname} server's api.
```
Comment (chiaradiamarcelo) [inline modules/hub/pages/autoconnect-mode.adoc]: same comments as in manual mode example

File: modules/hub/pages/hub-api-installation.adoc (hunk 5 above)
Comment (admd) [reply modules/hub/pages/hub-api-installation.adoc]: Yeah, we should.  :+1:

File: modules/hub/pages/hub-api-usage.adoc (hunk 4 above)
Comment (admd) [reply modules/hub/pages/hub-api-usage.adoc]: We will change this, once systemd solution is done.

**Comment by admd** [comment]
Thank you @chiaradiamarcelo  for the review. I have made most of the suggested changes.

**Comment by Loquacity** [comment]
> Thank you @chiaradiamarcelo for the review. I have made most of the suggested changes.

Thanks for all your hard work getting this written down. I've created a proposal for a new guide, which would incorporate this new content, in addition to the large scale tuning content we already have in the Salt Guide: https://github.com/uyuni-project/uyuni-docs/wiki/large-deployments-guide

Can you please review and let me know what you think?

**Comment by admd** [comment]
Hi @Loquacity ,
The proposal looks pretty promising to me, thank you for this.  It indeed makes a lot of sense to group this under large scale deployments umbrella. One thing  I should mention that we would like to have this documentation in beta3.
FYI @paususe  and @moio I

**Comment by Loquacity** [comment]
> Hi @Loquacity ,
> The proposal looks pretty promising to me, thank you for this. It indeed makes a lot of sense to group this under large scale deployments umbrella. One thing I should mention that we would like to have this documentation in beta3.
> FYI @paususe and @moio I

I'm not sure what the date of beta3 is with the shift in the schedule, so I'm not sure if that's possible or not.

@paususe can you please let me know what the new dates are, and whether or not you're happy for this to go ahead?

**Comment by paususe** [comment]
> @paususe can you please let me know what the new dates are, and whether or not you're happy for this to go ahead?

There are no changes in the beta 3 dates.

**Comment by Loquacity** [comment]
> > @paususe can you please let me know what the new dates are, and whether or not you're happy for this to go ahead?
> 
> There are no changes in the beta 3 dates.

So basically it's still two weeks away, with Easter during that period, so we're probably pushing it. Can we call it 4.1 instead?

**Comment by Loquacity** [comment]
> > > @paususe can you please let me know what the new dates are, and whether or not you're happy for this to go ahead?
> > 
> > 
> > There are no changes in the beta 3 dates.
> 
> So basically it's two weeks away, with Easter during that period, so we're probably pushing it. Can we call it RC1 instead?

I can have an outline with initial content up pretty quickly, but I think it's going to require a reasonable amount of back and forth to get the new content ready.

File: modules/hub/pages/hub-xmlrpc-api-intro.adoc [hunk 21]
```diff
@@ -0,0 +1,6 @@
+[[hub-xmlrpc-api]]
+= Hub XMLRPC API
+
+
+The Hub XMLRPC API helps to operate SUSE Manager or Uyuni infrastructures with one Server, called a Hub, managing several Servers.
+Servers are registered and can be managed from the Hub just like normal SUSE Manager or Uyuni clients. In addition to that, the Hub XMLRPC API provides a more convenient way to access all Servers' XMLRPC APIs from one only endpoint in the Hub.
```
Comment (moio) [inline modules/hub/pages/hub-xmlrpc-api-intro.adoc:6]: "In addition to that" seems out of place here - this is the introduction to the Hub XMLRPC API, and only that should be described.

File: modules/hub/pages/hub-api-requirements.adoc [hunk 22]
```diff
@@ -0,0 +1,8 @@
+[[hub-api-namespaces]]
+= Hub XMRPC API Requirements
+
+[square]
```
Comment (moio) [inline modules/hub/pages/hub-api-requirements.adoc:4]: What is this `[square]`?

File: modules/hub/pages/hub-api-requirements.adoc [hunk 23]
```diff
@@ -0,0 +1,8 @@
...
+* One central SUSE Manager/Uyuni Server (hereinafter: "Hub")
```
Comment (moio) [inline modules/hub/pages/hub-api-requirements.adoc:5]:
```suggestion
* One central {productname} Server (hereinafter: "Hub")
```

File: modules/hub/pages/hub-api-requirements.adoc [hunk 24]
```diff
@@ -0,0 +1,8 @@
...
+* Two or more peripheral Servers registered to the Hub as Salt clients
+* Any number of end clients registered to peripheral Servers
+You will need credentials to all Server XMLRPC APIs, including the Hub's.
```
Comment (moio) [inline modules/hub/pages/hub-api-requirements.adoc:8]:
```suggestion
 * credentials to all Server XMLRPC APIs, including the Hub's
```

File: modules/hub/pages/hub-api-usage.adoc (hunk 4 above)
Comment (moio) [reply modules/hub/pages/hub-api-usage.adoc:4]:
```suggestion
`hub-xmlrpc-api` is a daemon to be launched from the command line.
```

(it will change when it changes)

File: modules/hub/pages/hub-api-installation.adoc [hunk 25]
```diff
@@ -0,0 +1,21 @@
+[[hub-api-installation]]
+= Hub XMRPC API Installation
+
+
```
Comment (moio) [inline modules/hub/pages/hub-api-installation.adoc:4]:
```suggestion
```

File: modules/hub/pages/hub-api-installation.adoc [hunk 26]
```diff
@@ -0,0 +1,21 @@
...
+Install the package hub-xmlrpc-api, available from SUSE Manager 4.1 and Uyuni repos, either on the Hub itself or on a host that has access to all Servers' XMLRPC APIs.
```
Comment (moio) [inline modules/hub/pages/hub-api-installation.adoc:5]:
```suggestion
Install the `hub-xmlrpc-api` package, which is available from {productname} repositories, either on the Hub itself or on a host that has access to all Servers' XMLRPC APIs.
```

File: modules/hub/pages/hub-api-installation.adoc [hunk 27]
```diff
@@ -0,0 +1,21 @@
...
+
+Configuration of `hub-xmlrpc-api` is specified in a JSON file like the following:
```
Comment (moio) [inline modules/hub/pages/hub-api-installation.adoc:7]: Are there defaults for the values below? Ie. will it work if I set nothing at all?

File: modules/hub/pages/hub-api-usage.adoc [hunk 28]
```diff
@@ -0,0 +1,7 @@
...
+
+Once running, you can connect to the hub-xmlrpc-api at port 8888 via any XMLRPC compliant client libraries.
```
Comment (moio) [inline modules/hub/pages/hub-api-usage.adoc:6]: Please always use the proper name "Hub XMLRPC API" when referring to the component in general and use backticks (`hub-xmlrpc-api`) only when referring to the executable file (or the systemd unit, when it is implemented).

File: modules/hub/pages/hub-api-namespaces.adoc [hunk 29]
```diff
@@ -0,0 +1,26 @@
+[[hub-api-namespaces]]
+= Hub XMRPC API namespaces
+
+
+Hub supports 3 different namespaces.
```
Comment (moio) [inline modules/hub/pages/hub-api-namespaces.adoc:5]: Hub or Hub XMLRPC API?

What is a namespace?

File: modules/hub/pages/hub-api-namespaces.adoc [hunk 30]
```diff
@@ -0,0 +1,26 @@
...
+
+
+[[name-spaces]]
+[cols="0,1,1", options="header"]
+.Namespaces
+|===
+| Name                  | Description  | Example
+| hub                   | to target the Hub itself. | `client.hub.login(login, password)`
```
Comment (moio) [inline modules/hub/pages/hub-api-namespaces.adoc:13]: Please expand the descriptions a bit (if they become too long, you might even consider not using a table).

Eg. "in this namespace are some Hub-specific XMLRPC API endpoints, generally not available on peripheral Servers. For a list of those endpoints see..."

or

"this namespace redirects any call transparently to one specific Server and returns back any return value as if the Server's XMLRPC API endpoint was used directly. Use this namespace to operate on an individual Server (including any client or {productname} Proxy attached to it)"

File: modules/hub/pages/hub-api-authentication-modes.adoc [hunk 31]
```diff
@@ -0,0 +1,14 @@
+[[hub-api-authentication-modes]]
+= Authentication modes
+
+Hub supports 3 different authentication modes.
+
+[square]
+* `Manual mode` (default): user needs to provide API credentials for each Server explicitly
+* `Relay mode`: the same credentials used to authenticate against the Hub will be re-used to authenticate Servers. The list of Servers to connect to will still be provided by the user
```
Comment (moio) [inline modules/hub/pages/hub-api-authentication-modes.adoc:8]:
```suggestion
* `Relay mode`: the same credentials used to authenticate against the Hub will be also used to authenticate against Servers. The list of Servers to connect to will still be provided by the user
```

This is useful if...

File: modules/hub/pages/autoconnect-mode.adoc [hunk 32]
```diff
@@ -0,0 +1,53 @@
+[[manual-mode-example]]
+= Manual mode
```
Comment (moio) [inline modules/hub/pages/autoconnect-mode.adoc:2]: Should be autoconnect mode I guess

File: modules/hub/pages/manual-mode.adoc [hunk 33]
```diff
@@ -0,0 +1,58 @@
+[[manual-mode-example]]
+= Manual mode
+
+
+In the Manual mode the user needs to provide credentials for each {productname} server the user wants to connect to.
```
Comment (moio) [inline modules/hub/pages/manual-mode.adoc:5]:
```suggestion
In manual mode credentials have to be explicitly provided for each {productname} server the user wants to connect to.
```

File: modules/hub/pages/manual-mode.adoc [hunk 34]
```diff
@@ -0,0 +1,58 @@
...
+
+Here in this example, following is happening
```
Comment (moio) [inline modules/hub/pages/manual-mode.adoc:7]:
```suggestion
In the following example:
```

File: modules/hub/pages/manual-mode.adoc [hunk 35]
```diff
@@ -0,0 +1,58 @@
...
+
+[square]
+* User provides credentials for the hub to the login method, a `sessionkey` for the hub is returned
```
Comment (moio) [inline modules/hub/pages/manual-mode.adoc:10]:
```suggestion
* credentials for the Hub are passed to the `login` method, a session key for the Hub API is returned (`hubKey`)
```

File: modules/hub/pages/manual-mode.adoc [hunk 36]
```diff
@@ -0,0 +1,58 @@
...
+* Using the sessionkey from the previous step, `serverIds` for the servers attached to the hub are requested by calling `listServerIds` method
```
Comment (moio) [inline modules/hub/pages/manual-mode.adoc:11]:
```suggestion
* Using the session key from the previous step, {productname} Server IDs are obtained for all peripheral Servers attached to the Hub via the `hub.listServerIds` method
```

File: modules/hub/pages/manual-mode.adoc [hunk 37]
```diff
@@ -0,0 +1,58 @@
...
+* User provides the credentials for each {productname} server, she needs data from and calls the hub method `attachToServers`. Authentication part is done here.
```
Comment (moio) [inline modules/hub/pages/manual-mode.adoc:12]:
```suggestion
* Credentials for each {productname} Server are now provided to the `attachToServers` method. This performs authentications against each of the Servers' XMLRPC API endpoints.
```

File: modules/hub/pages/manual-mode.adoc [hunk 38]
```diff
@@ -0,0 +1,58 @@
...
+* A `multicast` call is performed on the specific servers where `serverId` is used to identify the specific sever. Behind the scenes, `system.list_system method` of each {productname} server's XMLRPC API method will be called to get the list of the clients registered with each {productname} server
```
Comment (moio) [inline modules/hub/pages/manual-mode.adoc:13]:
```suggestion
* A `multicast` call is performed on a set of Servers (`serverIds` contains all IDs of Servers to target). Behind the scenes, `system.list_system` is called on each {productname} Server's XMLRPC API
```

File: modules/hub/pages/manual-mode.adoc [hunk 39]
```diff
@@ -0,0 +1,58 @@
...
+* Hub will aggregate the results and return the response in form of a `map`. This map has two entries:
+- `Successful`: contains a list of responses for those clients where the call was successful.
+- `Failed`: contains a list of responses for those clients where the call failed.
+
+[NOTE]
+====
+If user wish to get information from only one {productname} server, then Hub API also provided a convenient `unicast` namespace. In this case, the response would not be a map, instead the response format would be exactly the same as if the user would directly call that {productname} server's API.
```
Comment (moio) [inline modules/hub/pages/manual-mode.adoc:20]:
```suggestion
If you want to call a method on just one {productname} Server, then Hub API also provides a more convenient `unicast` namespace. In this case, the response will be a single value and not a map, just as if the user called that {productname} server's API directly.
```

File: modules/hub/pages/manual-mode.adoc [hunk 40]
```diff
@@ -0,0 +1,58 @@
...
+serverIds = client.hub.listServerIds(hubKey)
+
+# authenticate those servers(each server with it's own credentials)
+usernames = ["admin" for s in serverIds]
+passwords = ["admin" for s in serverIds]
+client.hub.attachToServers(hubKey, serverIds, usernames, passwords)
+
+# perform the needed operation
+systemsPerServer = client.multicast.system.list_systems(hubKey, serverIds)
+successfulResponses = systemsPerServer["Successfull"]["Responses"]
```
Comment (moio) [inline modules/hub/pages/manual-mode.adoc:50]:
```suggestion
successfulResponses = systemsPerServer["Successful"]["Responses"]
```

Apparently the field is spelled correctly in `hub-xmlrpc-api`'s source code.

I would recommend to double check all examples, as they could confuse first-time users if they do not work as expected.

File: modules/hub/pages/manual-mode.adoc [hunk 41]
```diff
@@ -0,0 +1,58 @@
...
+failedResponses = systemsPerServer["Failed"]["Responses"]
+
+for system in itertools.chain.from_iterable(successfulResponses):
```
Comment (moio) [inline modules/hub/pages/manual-mode.adoc:53]: Is the use of `itertools.chain.from_iterable` really necessary here?

I am most likely missing something - what would make `for response in successfulResponses` not work here?

File: modules/hub/pages/relay-mode.adoc [hunk 42]
```diff
@@ -0,0 +1,57 @@
+[[relay-mode-example]]
+= Relay mode
+
+
+The Relay mode mode is a convenient method where the same credentials used to login into the Hub API are used to login into the APIs of the servers the user wants to work with. It is assumed the credentials are valid in each server and correspond to an admin user. After logging in into the Hub API the user still needs to call attachToServers method.
```
Comment (moio) [inline modules/hub/pages/relay-mode.adoc:5]:
```suggestion
The Relay authentication mode is a convenient method where the same credentials used to login into the Hub API are used to login into the APIs of the peripheral Servers the user wants to work with. It is thus assumed that the credentials are valid in each Server and correspond to a user with appropriate permissions. After login, it is still necessary to to call `attachToServers` method to define the subset of Servers that will be targeted in subsequent calls.
```

File: modules/hub/pages/autoconnect-mode.adoc [hunk 43]
```diff
@@ -0,0 +1,53 @@
...
+
+
+The autoconnect mode is a convenient method where the same credentials used to login into the Hub API are used to login into the APIs of the servers the user wants to work with.
```
Comment (moio) [inline modules/hub/pages/autoconnect-mode.adoc:5]:
```suggestion
The autoconnect mode is a convenience mechanism that, similarly to the relay mode, uses the same credentials to perform a login into the Hub and peripheral Server APIs.
```

File: modules/hub/pages/autoconnect-mode.adoc [hunk 44]
```diff
@@ -0,0 +1,53 @@
...
+It is assumed the credentials are valid in each server and correspond to an admin user. Difference from relay mode is that there is no need to call `attachToServers` method which also means that there is no need to call `listServerIds` method to get the `serverIds`.
```
Comment (moio) [inline modules/hub/pages/autoconnect-mode.adoc:6]:
```suggestion
Differently from relay mode, there is no need to use the `attachToServers` method, as it is assumed the user wants to connect to all peripheral Servers under his control, and the Hub XMLRPC API will in fact log in to them all at Hub login time.
```

File: modules/hub/pages/autoconnect-mode.adoc [hunk 45]
```diff
@@ -0,0 +1,53 @@
...
+This process will happen automatically when user will login into hub using `loginWithAutoconnectMode` method. Selected servers will be the servers where Hub user will have access though.
+
+Here in this example, following is happening
```
Comment (moio) [inline modules/hub/pages/autoconnect-mode.adoc:9]: See copy editing in manual mode

**Comment by paususe** [comment]
> I can have an outline with initial content up pretty quickly, but I think it's going to require a reasonable amount of back and forth to get the new content ready.

That's fine. Something is better than nothing.

File: modules/hub/pages/autoconnect-mode.adoc (hunk 44 above)
Comment (chiaradiamarcelo) [reply modules/hub/pages/autoconnect-mode.adoc:6]: while you are at it, can we change every occurrence of "serverIds" or derivative forms to "serverIDs" ? (Including example scripts)

**Comment by Loquacity** [comment]
This is now superseded by the new guide.
